File: CoinChangeTwo.java
// Time Complexity: O(m*n)
// Space Complexity: O(m*n)
public class CoinChangeTwo {
    public static void main(String[] args) {
        int[] coins = { 1, 2, 5 };
        int amount = 11;
        int result = coinChange(amount, coins);
        System.out.println("No of ways :: " + result);
    }

    /**
     * 
     * @param amount
     * @param coins
     * @return the number of ways that make up that amount
     */
    public static int coinChange(int amount, int[] coins) {
        int m = coins.length;
        int n = amount;
        int[][] dp = new int[m + 1][n + 1];
        for (int i = 0; i <= m; i++) {
            dp[i][0] = 1;
        }
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (j < coins[i - 1]) {
                    dp[i][j] = dp[i - 1][j];
                } else {
                    dp[i][j] = dp[i - 1][j] + dp[i][j - coins[i - 1]];
                }

            }
        }
        return dp[m][n];

    }
}

File: Coin_Change_Two.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        n=len(coins)
        
        if len(coins)==0:
            return -1
        
        if amount==0:
            return 1
        
        t=[[0 for _ in range(amount+1)]for _ in range(n+1)]
        
        import math
        
        for i in range(n+1):
            for j in range(amount+1):
                if i==0: 
                    t[i][j]=0
                if j==0:
                    t[i][j]=1
        
                    
        for i in range(1,n+1):
            for j in range(1,amount+1):
                if coins[i-1]<=j:
                    t[i][j]=t[i][j-coins[i-1]] + t[i-1][j]
                else:
                    t[i][j]=t[i-1][j]
        
        return t[n][amount]
        